File: backend/app/brands.py
"""ブランド・競合分析（Phase 2, docs/development_plan.md セクション 41）。

カタログをブランド単位で集計し、平均スコア・利益率・推定利益合計・競合水準・OEM 比率・
優勢な商流方向を決定論的に導く（原則: セクション 93）。競合水準は販売市場側の平均競合数
から段階化する。
"""

from __future__ import annotations

from statistics import mean

from . import opportunity_engine
from .catalog import CatalogEntry
from .oem import _is_oem_brand
from .schemas import BrandStat, TradeDirection

# 競合水準の境界（販売市場の平均競合数）。
_LOW_COMPETITION = 40
_MEDIUM_COMPETITION = 100


def _competition_level(avg_competitors: float) -> str:
    if avg_competitors <= _LOW_COMPETITION:
        return "low"
    if avg_competitors <= _MEDIUM_COMPETITION:
        return "medium"
    return "high"


def brand_analysis(entries: list[CatalogEntry]) -> list[BrandStat]:
    """ブランド別の集計を推定利益合計の降順で返す。"""
    by_brand: dict[str, list[CatalogEntry]] = {}
    for entry in entries:
        by_brand.setdefault(entry.brand, []).append(entry)

    stats: list[BrandStat] = []
    for brand, items in by_brand.items():
        evaluations = [opportunity_engine.evaluate(e).best for e in items]
        # 販売市場側の競合数（有望方向に応じて日本/中国を選ぶ）。
        sell_competitors = [
            (e.japan.competitors if best.direction == TradeDirection.CN_TO_JP else e.china.competitors)
            for e, best in zip(items, evaluations)
        ]
        export_count = sum(1 for best in evaluations if best.direction == TradeDirection.JP_TO_CN)
        import_count = len(evaluations) - export_count
        dominant: TradeDirection | None = None
        if export_count > import_count:
            dominant = TradeDirection.JP_TO_CN
        elif import_count > export_count:
            dominant = TradeDirection.CN_TO_JP

        avg_competitors = round(mean(sell_competitors))
        stats.append(
            BrandStat(
                brand=brand,
                product_count=len(items),
                avg_score=round(mean(best.score for best in evaluations)),
                avg_margin_rate=round(mean(best.economics.margin_rate for best in evaluations), 4),
                total_estimated_profit=sum(best.economics.estimated_profit for best in evaluations),
                avg_competitors=avg_competitors,
                competition_level=_competition_level(avg_competitors),
                oem_share=round(sum(1 for e in items if _is_oem_brand(e.brand)) / len(items), 4),
                dominant_direction=dominant,
            )
        )

    return sorted(stats, key=lambda s: s.total_estimated_profit, reverse=True)

File: src/lib/research/brands.ts
/**
 * ブランド・競合分析（Phase 2、docs/development_plan.md セクション 41）。
 *
 * backend/app/brands.py と同一の集計・段階化を移植し、API 未接続のモックでも同じ
 * ブランド別統計を返す（整合性の原則: セクション 93）。
 */

import { pyRound } from "./history";
import { isOemBrand } from "./oem";
import { evaluate } from "./opportunity-engine";
import type { BrandStat, CompetitionLevel, ProductCatalogEntry, TradeDirection } from "./types";

const LOW_COMPETITION = 40;
const MEDIUM_COMPETITION = 100;

function competitionLevel(avgCompetitors: number): CompetitionLevel {
  if (avgCompetitors <= LOW_COMPETITION) return "low";
  if (avgCompetitors <= MEDIUM_COMPETITION) return "medium";
  return "high";
}

function mean(values: number[]): number {
  return values.length ? values.reduce((a, b) => a + b, 0) / values.length : 0;
}

/** Python の round(value, digits)（偶数丸め）に一致させる。 */
function round(value: number, digits = 0): number {
  const factor = 10 ** digits;
  return pyRound(value * factor) / factor;
}

/** ブランド別の集計を推定利益合計の降順で返す。 */
export function brandAnalysis(entries: ProductCatalogEntry[]): BrandStat[] {
  const byBrand = new Map<string, ProductCatalogEntry[]>();
  for (const entry of entries) {
    const list = byBrand.get(entry.brand) ?? [];
    list.push(entry);
    byBrand.set(entry.brand, list);
  }

  const stats: BrandStat[] = [];
  for (const [brand, items] of byBrand) {
    const evaluations = items.map((e) => evaluate(e).best);
    const sellCompetitors = items.map((e, i) =>
      evaluations[i].direction === "CN_TO_JP" ? e.japan.competitors : e.china.competitors,
    );
    const exportCount = evaluations.filter((b) => b.direction === "JP_TO_CN").length;
    const importCount = evaluations.length - exportCount;
    let dominant: TradeDirection | null = null;
    if (exportCount > importCount) dominant = "JP_TO_CN";
    else if (importCount > exportCount) dominant = "CN_TO_JP";

    const avgCompetitors = round(mean(sellCompetitors));
    stats.push({
      brand,
      productCount: items.length,
      avgScore: round(mean(evaluations.map((b) => b.score))),
      avgMarginRate: round(mean(evaluations.map((b) => b.economics.marginRate)), 4),
      totalEstimatedProfit: evaluations.reduce((sum, b) => sum + b.economics.estimatedProfit, 0),
      avgCompetitors,
      competitionLevel: competitionLevel(avgCompetitors),
      oemShare: round(items.filter((e) => isOemBrand(e.brand)).length / items.length, 4),
      dominantDirection: dominant,
    });
  }

  return stats.sort((a, b) => b.totalEstimatedProfit - a.totalEstimatedProfit);
}
